gh-94518: [_posixsubprocess] Replace variable validity flags with reserved values

arhadthedev opened the pull request:
Currently, `_posixsubprocess.c` uses group and user ids together with flags of whether the corrsponding variables were initialized. However such an implicit "either initialized or look somewhere else" confuses both a reader (another mental connection to constantly track between functions) and a compiler (warnings on potentially uninitialized variables). Instead, we can utilize a special group/user id as a flag value defined by POSIX but uses nowhere else. Namely:

- `gid`: `call_setgid = False` → `gid = -1`
- `uid`: `call_setuid = False` → `uid = -1`
- `groups`: `call_setgroups = False` → `groups = NULL` (obtained with `(groups_list != Py_None) ? groups : NULL`)

This PR is required for gh-94519.

---

PS.: @tiran I applied your suggestion from gh-94519 as this, separate PR to keep that one limited in scope and reviewable.

<!-- gh-issue-number: gh-94518 -->
* Issue: gh-94518
<!-- /gh-issue-number -->

arhadthedev committed:
Replace call_setgid/call_setuid in _posixsubprocess with gid/uid==-1

arhadthedev committed:
Merge branch 'main' into posixsubprocess-no-callsetxid

arhadthedev committed:
Add a NEWS entry

arhadthedev committed:
Address Gregory's review in gh-94519

gpshead commented on Modules/_posixsubprocess.c:
```diff
@@ -737,10 +737,7 @@ do_fork_exec(char *const exec_array[],
-        assert(!call_setuid);
-        assert(!call_setgid);
-        assert(!call_setgroups);
+        /* Checked by our caller; verify this in debug builds. */
```
keep equivalent assertions in here:

```
assert(uid == (uid_t)-1);
assert(gid == (gid_t)-1);
assert(groups_size < 0);
```

gpshead commented on Modules/_posixsubprocess.c:
```diff
@@ -724,9 +724,9 @@ do_fork_exec(char *const exec_array[],
-             int call_setgroups, size_t groups_size, const gid_t *groups,
-             int call_setuid, uid_t uid, int child_umask,
+             gid_t gid,
+             size_t groups_size, const gid_t *groups,
```
change groups_size to a Py_ssize_t as that is what we pass in. We can then use it being < 0 to mean the former `!call_setgroups`.  A call to C's `setgroups(0, NULL)` via an empty list of groups has meaning.

gpshead commented on Modules/_posixsubprocess.c:
```diff
@@ -518,9 +518,9 @@ child_exec(char *const exec_array[],
-           int call_setgroups, size_t groups_size, const gid_t *groups,
-           int call_setuid, uid_t uid, int child_umask,
+           gid_t gid,
+           size_t groups_size, const gid_t *groups,
```
Use Py_ssize_t for groups_size so that < 0 can mean "don't call setgroups".

gpshead commented on Modules/_posixsubprocess.c:
```diff
@@ -619,17 +619,17 @@ child_exec(char *const exec_array[],
 
 #ifdef HAVE_SETGROUPS
-    if (call_setgroups)
+    if (groups)
```
`if (groups_size >= 0)`

`setgroups(0, NULL)` is valid.

gpshead commented on Modules/_posixsubprocess.c:
```diff
@@ -951,34 +946,31 @@ subprocess_fork_exec(PyObject *module, PyObject *args)
             }
```
Change

```
if ((groups = PyMem_RawMalloc(num_groups * sizeof(gid_t))) == NULL) {
```

to

```
if (num_groups > 0 && (groups = PyMem_RawMalloc(num_groups * sizeof(gid_t))) == NULL) {
```

gpshead commented on Modules/_posixsubprocess.c:
```diff
@@ -1002,7 +994,7 @@ subprocess_fork_exec(PyObject *module, PyObject *args)
     sigset_t old_sigs;
     if (preexec_fn == Py_None && allow_vfork &&
-        !call_setuid && !call_setgid && !call_setgroups) {
+        uid == (uid_t)-1 && gid == (gid_t)-1 && groups_list == Py_None) {
```
for consistency, use `num_groups < 0` here and below instead of the groups_list Py_None check.

gpshead reviewed (changes requested):
I like this PR :)

arhadthedev committed:
Restore assertions in do_fork_exec()

arhadthedev committed:
Check num_groups for emptiness, not groups_list

`num_groups < 0` already implies `groups_list == Py_None` because
`num_groups = PySequence_Size(groups_list)` does this check for us.

Also, it guarantees that `groups_list == Py_None` <=> `groups == NULL`;
it allows to replace `groups_list != Py_None ? groups : NULL` with just
`groups` in a call of `do_fork_exec()`.

arhadthedev committed:
Tighten scope of relevant short lived variables

arhadthedev committed:
Use negative groups_size for "don't call setgroups"

When we query `groups_list` size using `PySequence_Size()`, we get -1
for non-lists. It returns an ability to pass an empty list to
`setgroups()` that is valid but treated as a no-op:

> If `size` is zero, `list` is not modified, but the total number of
> supplementary group IDs for the process is returned. This allows the
> caller to determine the size of a dynamically allocated list to be
> used in a further call to getgroups().
>
> - <https://linux.die.net/man/2/setgroups>

arhadthedev committed:
Fix `PermissionError: [Errno 1] Operation not permitted` exception

arhadthedev committed:
Merge branch 'main' into posixsubprocess-no-callsetxid

arhadthedev committed:
Merge branch 'main' into posixsubprocess-no-callsetxid

arhadthedev committed:
Merge branch 'main' into posixsubprocess-no-callsetxid

arhadthedev committed:
Merge branch 'main' into posixsubprocess-no-callsetxid

arhadthedev committed:
Merge branch 'main' into posixsubprocess-no-callsetxid

gpshead reviewed (approved):
thanks for doing this!
